### PR TITLE
fix --source option not working for outdated and update commands

### PR DESCRIPTION
**Pull request by colby-swandale**
Relates to #4881

This fixes 2 related problems for the `outdated` and `update` commands in Bundler. Both commands have an option called `source` where you can specify a rubygems server not in the Gemfile for updated/outdated to use.

Currently the option for both commands is not working. See example:
```
source 'https://rubygems.org'

gem 'rack', '~> 1.6.5'
```
```
› bundle outdated --source https://mytestserver.com
Fetching gem metadata from https://rubygems.org/..............
Fetching version metadata from https://rubygems.org/.
Resolving dependencies...

Outdated gems included in the bundle:
  * rack (newest 2.0.1, installed 1.6.5, requested ~> 1.6.5) in groups "default"
```
```
› bundle update --source https://mytestserver.com
Fetching gem metadata from https://rubygems.org/..............
Fetching version metadata from https://rubygems.org/.
Resolving dependencies...
Using rack 1.6.5
Using bundler 1.14.0
Bundle updated!
Gems in the group test were not installed.
```

For both examples it should have been fetching the index/updating the gem from https://mytestserver.com.

This fix looks like part of a larger design issue, [specifically here](https://github.com/bundler/bundler/blob/master/lib/bundler.rb#L124).  Maybe work on that in the future.

I'm also not fully convinced of my patch in `dsl.rb` is the best solution for now, so i'm very open to suggestions.

### CODE REVIEW

File: spec/commands/update_spec.rb [hunk 1]
```diff
@@ -635,6 +635,26 @@
...
+    it 'updates gems using the given source' do
+      gemfile <<-G
+        source "file://#{gem_repo1}"
+        gem 'rack'
+      G
+
+      bundle "install"
+
+      bundle "update --source file://#{gem_repo2}"
+      expect(out).to include "Fetching source index from file:#{gem_repo2}/"
```
Comment (colby-swandale) [inline spec/commands/update_spec.rb]: I realise a better test for this would be to actually check `the_bundle` instead

File: lib/bundler/dsl.rb [hunk 2]
```diff
@@ -197,6 +197,7 @@ def github(repo, options = {})
     end
 
     def to_definition(lockfile, unlock)
+      unlock[:sources].each { |s| @sources.add_rubygems_remote(s) } if unlock.is_a?(Hash) && unlock[:sources]
```
Comment (segiddins) [inline lib/bundler/dsl.rb]: Maybe this belongs in the definition initializer?

File: lib/bundler/definition.rb [hunk 3]
```diff
@@ -113,6 +114,10 @@ def initialize(lockfile, dependencies, sources, unlock, ruby_version = nil, opti
         @unlock[:gems] = @locked_specs.for(eager_unlock).map(&:name)
       end
 
+      if @unlock[:sources]
+        @unlock[:sources].each {|s| @sources.add_rubygems_remote(s) } if @unlock.is_a?(Hash) && @unlock[:sources]
```
Comment (olleolleolle) [inline lib/bundler/definition.rb]: You already check `if unlock[:sources]` in the surrounding `if` block. The last bit of the suffix `if` can be removed, right?

File: spec/commands/outdated_spec.rb [hunk 4]
```diff
@@ -595,6 +595,25 @@ def test_group_option(group = nil, gems_list_size = 1)
     it_behaves_like "patch version updates are detected"
   end
 
+  context "with --soruce option" do
```
Comment (olleolleolle) [inline spec/commands/outdated_spec.rb]: Typo: `--soruce` => `--source`

File: lib/bundler/definition.rb [hunk 5]
```diff
@@ -113,6 +114,10 @@ def initialize(lockfile, dependencies, sources, unlock, ruby_version = nil, opti
         @unlock[:gems] = @locked_specs.for(eager_unlock).map(&:name)
       end
 
+      if @unlock.is_a?(Hash) && @unlock[:sources]
```
Comment (olleolleolle) [inline lib/bundler/definition.rb]: Come to think of this line, line 112 with its `unless @unlock[:lock_shared_dependencies]` would have already failed if the `:[]` method was not available.

**Comment by colby-swandale** [comment]
I have been asked why i closed this. It's because i misunderstood the behaviour of the `--source` option and this PR would have introduce undesirable behaviour.

**Comment by mswiszcz** [comment]
@colby-swandale could you recreate this and add additional option as proposed in #4881 ? like `--remote` - it would be really useful for me as well

**Pull request closed without merging**